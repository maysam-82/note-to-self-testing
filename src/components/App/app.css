.button-submit {
	width: 30%;
}

.form-control-input {
	width: calc(70% - 1rem) !important;
}
