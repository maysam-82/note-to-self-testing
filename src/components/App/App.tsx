import React, { Component } from 'react';
import { Form, FormControl, Button } from 'react-bootstrap';
import Note from '../Note';
import {
	saveItemToLocalStorage,
	getItemFromLocalStorage,
	removeItemFromLocalStorage,
} from '../../Utilities/storage';
import './app.css';
export interface IAppState {
	notes: string[];
	note: string;
	selectedNoteId: string;
}

class App extends Component<{}, IAppState> {
	constructor(props: {}) {
		super(props);

		this.state = {
			note: '',
			notes: [],
			selectedNoteId: '',
		};
	}

	componentDidMount() {
		this.getNotes();
	}

	getNotes = () => {
		const savedNotes = getItemFromLocalStorage<string[]>('notes');
		if (savedNotes) {
			this.setState((prevState) => ({
				notes: [...prevState.notes, ...savedNotes],
			}));
		} else {
			return;
		}
	};

	handleOnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const note = event.target.value;
		this.setState((prevState) => ({ note }));
	};

	handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
		event.preventDefault();
		const { note, selectedNoteId } = this.state;
		if (note) {
			if (selectedNoteId) {
				const noteId = parseInt(selectedNoteId);
				const clonedNotes = [...this.state.notes];
				clonedNotes[noteId] = this.state.note;
				this.setState({ notes: clonedNotes, selectedNoteId: '' }, () =>
					this.saveNote()
				);
			} else {
				this.setState(
					(prevState) => ({
						notes: [...prevState.notes, this.state.note],
					}),
					() => this.saveNote()
				);
			}
		}
	};
	saveNote = () => {
		saveItemToLocalStorage<string[]>('notes', this.state.notes);
		this.setState({ note: '' });
	};
	handleDeleteNote = (noteId: number) => {
		const filteredNotes = this.state.notes.filter(
			(_, index) => index !== noteId
		);
		this.setState(
			{
				notes: filteredNotes,
			},
			() => {
				removeItemFromLocalStorage('notes');
				saveItemToLocalStorage<string[]>('notes', this.state.notes);
			}
		);
	};
	clearNotes = () => {
		this.setState(
			{
				notes: [],
			},
			() => {
				removeItemFromLocalStorage('notes');
			}
		);
	};
	handleEditNote = (noteId: number) => {
		const selectedNote = this.state.notes.find((_, index) => index === noteId);
		if (selectedNote) {
			this.setState({ note: selectedNote, selectedNoteId: noteId.toString() });
		}
	};
	renderNotes = () => {
		return this.state.notes.map((note, index) => (
			<Note
				key={index}
				note={note}
				id={index}
				handleDeleteNote={this.handleDeleteNote}
				handleEditNote={this.handleEditNote}
			/>
		));
	};
	render() {
		const { note, notes, selectedNoteId } = this.state;
		return (
			<div className="container">
				<h1 className="mt-3 mb-2 text-center">Note To Self</h1>
				<div className="text-center">{this.renderNotes()}</div>
				<div
					className={`pt-2 pb-2 mt-2 ${notes.length > 0 ? 'border-top' : ''}`}
				>
					<Form
						className="d-flex justify-content-between"
						onSubmit={(event: React.FormEvent<HTMLFormElement>) =>
							this.handleSubmit(event)
						}
					>
						<FormControl
							onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
								this.handleOnChange(event)
							}
							value={note}
							className="form-control-input"
						/>
						<Button className="button-submit" type="submit">
							{!selectedNoteId ? 'Submit' : 'Edit'}
						</Button>
					</Form>
					<div className="pb-2 pt-2 d-flex justify-content-center button-clear-notes">
						<Button onClick={this.clearNotes} variant="danger" block>
							Clear Notes
						</Button>
					</div>
				</div>
			</div>
		);
	}
}

export default App;
