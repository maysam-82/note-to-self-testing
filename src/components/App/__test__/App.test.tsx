import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import App, { IAppState } from '../App';

describe('App component', () => {
	const component = mount<App, {}, IAppState>(<App />);
	const note = 'sample note';
	it('should render App component', () => {
		expect(component).toMatchSnapshot();
	});
	it('should initialize the `state` with an empty `notes`', () => {
		expect(component.state().notes).toEqual([]);
	});
	it('should initialize the `state` with an empty `note`', () => {
		expect(component.state().note).toEqual('');
	});
	it('should initialize the `state` with an empty `note`', () => {
		expect(component.state().selectedNoteId).toEqual('');
	});
	it('should have an h1 with text `Note To Self', () => {
		expect(component.find('h1').text()).toEqual('Note To Self');
	});

	describe('If value of input is changed', () => {
		beforeEach(() => {
			component
				.find('FormControl')
				.simulate('change', { target: { value: note } });
		});

		it('should update `note` value in `state`', () => {
			expect(component.state().note).toEqual(note);
		});
		describe('If `Submit` button is clicked', () => {
			beforeEach(() => {
				component.find('Form').simulate('submit');
			});

			afterEach(() => {
				component.find('Button').at(1).simulate('click');
			});
			it('should add value of `note` in the `state` to `notes`', () => {
				expect(component.state().notes[0]).toEqual(note);
			});
			describe('App is re-mounting', () => {
				let remountComponent: ReactWrapper<{}, IAppState, App>;
				beforeEach(() => {
					remountComponent = mount(<App />);
				});
				it('should read notes from `notes` in the `localStorage`', () => {
					expect(remountComponent.state().notes).toEqual([note]);
				});
			});
		});
	});
	describe('When `Clear Notes` is clicked', () => {
		beforeEach(() => {
			component.setState({ notes: [note] });
			// component.instance().clearNotes();
			component.find('Button').at(1).simulate('click');
		});
		it('should remove all notes from `notes` in `state', () => {
			expect(component.state().notes).toEqual([]);
		});
	});
	describe('If Delete icon is clicked', () => {
		component.setState({ notes: [note] });
		component.find('.delete').at(0).simulate('click');
		it('should remove a `note` from `notes`', () => {
			expect(component.state().notes).toEqual([]);
		});
	});
});
