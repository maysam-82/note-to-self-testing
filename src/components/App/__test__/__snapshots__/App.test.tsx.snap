// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render App component 1`] = `
<App>
  <div
    className="container"
  >
    <h1
      className="mt-3 mb-2 text-center"
    >
      Note To Self
    </h1>
    <div
      className="text-center"
    />
    <div
      className="pt-2 pb-2 mt-2 "
    >
      <Form
        className="d-flex justify-content-between"
        inline={false}
        onSubmit={[Function]}
      >
        <form
          className="d-flex justify-content-between"
          onSubmit={[Function]}
        >
          <FormControl
            className="form-control-input"
            onChange={[Function]}
            value=""
          >
            <input
              className="form-control-input form-control"
              onChange={[Function]}
              value=""
            />
          </FormControl>
          <Button
            active={false}
            className="button-submit"
            disabled={false}
            type="submit"
            variant="primary"
          >
            <button
              className="button-submit btn btn-primary"
              disabled={false}
              type="submit"
            >
              Submit
            </button>
          </Button>
        </form>
      </Form>
      <div
        className="pb-2 pt-2 d-flex justify-content-center button-clear-notes"
      >
        <Button
          active={false}
          block={true}
          disabled={false}
          onClick={[Function]}
          type="button"
          variant="danger"
        >
          <button
            className="btn btn-danger btn-block"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Clear Notes
          </button>
        </Button>
      </div>
    </div>
  </div>
</App>
`;
